skbio/metadata/_interval.py

```diff
@@ -0,0 +1,465 @@
+        """ Fetches Interval objects based on query interval"""
+        _assert_valid_interval(interval)
+        start, end = interval
+        invs = self._intervals.find(start, end)
+        return invs
+
+    def _query_attribute(self, intervals, metadata):
+        """ Fetches Interval objects based on query attributes"""
+        if metadata is None:
+            return []
+
+        for inv in intervals:
+            for (key, value) in metadata.items():
+                if inv.metadata[key] != value:
+                    continue
+                yield inv
+
+    @experimental(as_of='0.4.2-dev')
+    def query(self, intervals=None, boundaries=None, metadata=None):
+        """ Looks up Interval objects with the intervals, boundaries and keywords.
+
+        Parameters
+        ----------
+        intervals : iterable of tuple of ints
+            A list of intervals associated with the Interval object.
+        boundaries : iterable of tuple of bool
+            A list of boundaries associated with the Interval object.
+        metadata : dict
+            A dictionary of key word attributes associated with the
+            Interval object.
+
+        Returns
+        -------
+        generator, Interval
+            A generator of Interval objects satisfying the search criteria.
+
+        Examples
+        --------
+        >>> from skbio.metadata import IntervalMetadata
+        >>> interval_metadata = IntervalMetadata()
+        >>> interval_metadata.add(intervals=[(0, 2)],
+        ...                       boundaries=None, metadata={'name': 'sagA'})
+        >>> interval_metadata.add(intervals=[(40, 70)],
+        ...                       boundaries=None, metadata={'name': 'sagB'})
+        >>> list(interval_metadata.query(intervals=[(1, 2)]))
+        [<Interval: start=0, end=2, 1 contiguous interval, 1 metadata keys, \
+dropped=False>]
+
+        Note
+        ----
+        There are two types of queries to perform
+        1. Query by interval.
+        2. Query by key/val pair (i.e. 'gene':'sagA').
+
+        If no intervals are specified, then only metadata will be searched for.
+        If no metadata is specified, then only intervals will be searched for.
+        Otherwise, the search will return the intersection of the two results.
+        """
+        if self._is_stale_tree:
+            self._rebuild_tree(self._metadata)
+            self._is_stale_tree = False
+
+        # empty iterator
+        def empty():
+            return
+            yield
+        seen = set()
+        invs = empty()
+        if intervals is None and metadata is None:
+            return
+            yield
+        # only metadata specified
+        elif intervals is None and metadata is not None:
+            invs = self._query_attribute(self._metadata,
+                                         metadata)
+            for q in invs:
+                if q._hash() not in seen:
+                    seen.add(q._hash())
```

Comment:
I don't think this hash is doing what you think - it's actually hashing the hash of an ``Interval`` object. 

If two ``Interval`` objects hash to the same number, but are not equal to one another then ``seen`` can't differentiate them, because it can only compare the hashes, not ``Interval`` objects. If ``Interval`` defined ``__hash__``, this would be avoided, but it can't because ``Interval`` is mutable. 

So, the solution here is to add ``id(q)`` to the set because the memory addresses will be unique and can't collide (while the hashes will eventually collide). 

See [here](https://docs.python.org/3/reference/datamodel.html#object.__hash__) for more detail on Python's ``__hash__`` and ``__eq__`` protocol.